src/sources/util/decoding/mod.rs

```diff
@@ -0,0 +1,196 @@
+
+impl DecoderError for LinesCodecError {}
+
+impl From<std::io::Error> for Error {
+    fn from(error: std::io::Error) -> Self {
+        Self {
+            error: std::sync::Arc::new(error),
+        }
+    }
+}
+
+impl From<LinesCodecError> for Error {
+    fn from(error: LinesCodecError) -> Self {
+        Self {
+            error: std::sync::Arc::new(error),
+        }
+    }
+}
+
+pub struct Decoder {
+    framer: Box<dyn tokio_util::codec::Decoder<Item = Bytes, Error = Error> + Send + Sync>,
+    parser: Box<dyn Parser + Send + Sync>,
+}
+
+impl Decoder {
+    pub fn new(
+        framer: Box<
+            dyn tokio_util::codec::Decoder<Item = Bytes, Error = Error> + Send + Sync + 'static,
+        >,
+        parser: Box<dyn Parser + Send + Sync + 'static>,
+    ) -> Self {
+        Self { framer, parser }
+    }
+}
+
+impl tokio_util::codec::Decoder for Decoder {
+    type Item = (Event, usize);
+    type Error = self::Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.framer.decode(buf).map(|result| {
+            result.and_then(|frame| {
+                let byte_size = frame.len();
+                match self.parser.parse(frame) {
+                    Ok(event) => Some((event, byte_size)),
+                    Err(error) => {
+                        emit!(DecoderParseFailed { error });
+                        None
+                    }
+                }
+            })
+        })
+    }
+
+    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.framer.decode_eof(buf).map(|result| {
+            result.and_then(|frame| {
+                let byte_size = frame.len();
+                match self.parser.parse(frame) {
+                    Ok(event) => Some((event, byte_size)),
+                    Err(error) => {
+                        emit!(DecoderParseFailed { error });
+                        None
+                    }
+                }
+            })
+        })
+    }
+}
+
+pub struct BytesDecoder<Error: From<std::io::Error> + Into<self::Error>, Item: Into<Bytes> = Bytes>
+{
+    decoder:
+        Box<dyn tokio_util::codec::Decoder<Item = Item, Error = Error> + Send + Sync + 'static>,
+}
+
+impl<Error: From<std::io::Error> + Into<self::Error>, Item: Into<Bytes>> BytesDecoder<Error, Item> {
+    pub fn new(
+        decoder: impl tokio_util::codec::Decoder<Item = Item, Error = Error> + Send + Sync + 'static,
+    ) -> Self {
+        Self {
+            decoder: Box::new(decoder),
+        }
+    }
+}
+
+impl<Error: From<std::io::Error> + Into<self::Error>, Item: Into<Bytes>> tokio_util::codec::Decoder
+    for BytesDecoder<Error, Item>
+{
+    type Item = Bytes;
+    type Error = self::Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.decoder
+            .decode(buf)
+            .map(|result| result.map(|frame| frame.into()))
```

Comment:
I'm a little bit concerned that this whole `BytesDecoder` wrapper thing will obfuscate the `Into<Bytes>` operation, which could involve undesirable and performance-impacting copying. We should try to work directly with `Bytes` where possible, but it should be explicitly handled when there's a source or decoder that doesn't.

Reply:
`BytesCodec` exists to reuse codecs where the output is not `Item = Bytes, Error = decoding::Error`, but can be converted trivially.

Such as `LengthDelimitedCodec` (which has `Item = BytesMut, Error = io::Error`), `LinesCodec` (which has `Item = String, Error = LinesCodecError`) and `OctetCountingDecoder` (which has `Output = Bytes, Error = LinesCodecError`).

The conversion between `BytesMut` or `String` to `Bytes` is zero-copy, and as far as I can see `Bytes` doesn't implement the `From` trait when it wouldn't be zero-copy. However, that wouldn't prevent us from implementing a type that implements the `From` trait for `Bytes` in a possibly expensive fashion.

> We should try to work directly with Bytes where possible, but it should be explicitly handled when there's a source or decoder that doesn't.

How would you approach this - would you like to define new types e.g. `LengthDelimitedBytesCodec` for `LengthDelimitedCodec` that internally goes from `String` to `Bytes`? I'm not sure if that would obfuscate the conversion less since you still need to know if the `into` implementation is zero-copy

Reply:
I removed `BytesCodec` and specified the `Item` type for all our decoders explicitly, also see https://github.com/timberio/vector/pull/8236#discussion_r689127860 for some more context